### DISCUSSION TITLE
CI/CD

### DISCUSSION THREAD

**Discussion by gyungwan** [24tngus/CS_STUDY#19] [CS]
매번 개발자가 코드를 수정하고 빌드와 테스트를 하고 배포까지 한다면  상당히 많은 시간이 소요됩니다. 하지만 git에 코드를 올리는 것만으로도 누군가가 빌드와 테스트, 배포까지 해준다면, 쓸데없는 시간을 단축시키고 개발에 더 많은 시간을 투자할 수 있을겁니다.

## **CI 란?**

CI는 간단히 요약하자면 `빌드/테스트 자동화 과정` 과정입니다. CI는 개발자를 위한 자동화 프로세스인 `지속적인 통합(Continuous Integration)`을 의미합니다. CI를 성공적으로 구현할 경우 애플리케이션에 대한 새로운 코드 변경 사항이 정기적으로 빌드 및 테스트되어 공유 리포지토리에 통합되므로 여러 명의 개발자가 동시에 애플리케이션 개발과 관련된 코드 작업을 할 경우 서로 충돌할 수 있는 문제를 해결할 수 있습니다.

지속적 통합의 실행은 소스/버전 관리 시스템에 대한 변경 사항을 정기적으로 커밋하여 모든 사람에게 동일 작업 기반을 제공하는 것으로 시작합니다. **커밋할 때마다 빌드와 일련의 자동 테스트가 이루어져 동작을 확인하고 변경으로 인해 문제가 생기는 부분이 없도록 보장**합니다. 지속적 통합은 그 자체로 유익하지만 CI/CD 파이프라인을 구현하기 위한 첫 번째 단계이기도 합니다.

## **CD 란?**

`CD`는 간단히 말하면 배포 자동화 과정입니다. CD는 `지속적인 서비스 제공(Continuous Delivery)` 또는 `지속적인 배포(Continuous Deployment)`를 의미하며 이 두 용어는 상호 교환적으로 사용됩니다. 두 가지 의미 모두 파이프라인의 추가 단계에 대한 자동화를 뜻하지만 때로는 얼마나 많은 자동화가 이루어지고 있는지를 설명하기 위해 별도로 사용되기도 합니다.

`지속적 배포`는 빌드, 테스트 및 배포 단계를 자동화하는 DevOps 방식을 논리적 극한까지 끌어 올립니다. 코드 변경이 파이프라인의 이전 단계를 모두 성공적으로 통과하면 수동 개입 없이 해당 변경 사항이 프로덕션에 자동으로 배포됩니다. 지속적 배포를 채택하면 품질 저하 없이 최대한 빨리 사용자에게 새로운 기능을 제공할 수 있습니다.

`지속적 배포`는 또한 입증된 지속적 통합 및 지속적인 전달 단계를 기반으로 합니다. **간단한 코드 변경이 정기적으로 마스터에 커밋되고, 자동화된 빌드 및 테스트 프로세스를 거치며 다양한 사전 프로덕션 환경으로 승격되며, 문제가 발견되지 않으면 최종적으로 배포**됩니다. 강력하고 신뢰할 수 있는 자동화 배포 파이프라인을 구축하면 하루에도 여러 번 이루어지는 릴리스가 특별하지 않은 일상이 됩니다.

![image](https://github.com/24tngus/CS_STUDY/assets/113571059/921b8609-ed4a-46e1-bf89-8a43fa659727)

#### CI/CD를 적용하기 전의 고전적인 코드 통합 과정을 생각해봅시다.

1. 개발자들이 개발하여 코드를 수정합니다.

2. 각자의 feature 브랜치에 코드를 push합니다. (but, 어느 한 부분에서 에러가 났지만 개발자들은 눈치채지 못합니다.)

3. 각자의 코드를 git에 올리고 통합(Intergration)합니다.

4. 에러가 발생했지만 어느 부분에서 에러가 났는지 모르므로 다시 어디부분에 에러가 있는지 디버깅하고 코드를 수정합니다.

5. (1) ~ (4)의 과정을 반복합니다.

6. 많은 시간을 할애하여 에러가 해결되었으면 배포를 시작합니다. 하지만 배포과정 또한, 개발자가 직접 배포과정을 거치므로 많은 시간을 소요합니다.

 

코드의 양이 적다면 조금만 시간을 할애해도 에러를 찾아낼 수 있지만, 코드의 양이 많다면 에러 추적이 안되므로 어마어마한 양의 디버깅 과정을 마주하게 될 수도 있습니다.

 

#### CI/CD를 적용 후의 과정

(이 과정은 하나의 예시일 뿐 다르게 변경할 수 있습니다. )

 

1. 개발자들이 개발하여 feature브랜치에 코드를 push합니다.

2. git push를 통해 Trigger되어 CI서버에서 알아서 Build, Test, Lint를 실행하고 결과를 전송합니다.

3. 개발자들은 결과를 전송받고 에러가 난 부분이 있다면 에러부분을 수정하고 코드를 master 브랜치에 merge합니다.

4. master 브랜치에 코드를 merge하고 Build, Test가 정상적으로 수행이 되었다면 CI서버에서 알아서 Deploy 과정을 수행합니다.

**Comment by berryberries** [comment]
확인했습니다.

**Comment by jindream6128** [comment]
멘토님 Comment)
우테코 팀 프로젝트에선 CI/CD가 필수 사항이라 모든 팀들이 구축하여 사용하고 있는데요, 가장 많이 쓰이는 건 Jenkins와 GitHub Actions 입니다. 요즈음은 GitHub Actions로 많이 쓰이는 추세라고 합니다. 세팅이 간단하기도 하고 GitHub와 통합이 잘 되어있거든요. 개발자가 직접 빌드를 실행하고 배포도 직접하는 번거로운 과정을 자동화하는 과정인 만큼 이를 직접 경험해본다면 정말 좋을 듯 합니다. (처음 시작한다면 GitHub Actions가 간단하기 때문에 추천드립니다)

**Comment by jindream6128** [comment]
확인 하였습니다 !

**Comment by sujin-park0607** [comment]
확인했습니다! 저는 CI는 gitlab이랑 githubActions를 써보고 CD는 CodeDeploy를 사용해서 구현해봤습니다! 생각보다 어렵더라구요 그래도 배운점이 많아서 다른분들도 배포자동화 만드실때 Actions,S3,CodeDeploy로 구현해보시면 좋을것같아요

**Comment by solo5star** [comment]
통합과 배포가 지속적으로 이루어지는 그 경험은 꼭 한번 해보면 정말 좋을 것 같습니다~ 프로젝트를 하실 때 간단한 CI/CD라도 해보시는 걸 추천드려요! 😁

**Comment by 24tngus** [comment]
확인했습니다! 실무에서 CI/CD 단어를 많이 사용하는데, 자세히 파악하게 되는 기회였습니다.

**Comment by jindream6128** [comment]
레퍼지토리 업로드 완료했습니다.
